File: elixir/2015/day06b/Program.exs
#!/usr/bin/env elixir

defmodule Program do
  def main(args) do
    case args do
      [filename] ->
        case process(filename) do
          {:ok, result} ->
            IO.puts("result = #{result}")
          {:error, reason} ->
            IO.puts(:stderr, "error while processing file `#{filename}': #{reason}")
            System.halt(1)
        end
      _ ->
        usage()
    end
  end

  defp usage do
    IO.puts(:stderr, "usage: Program.exs <input file>")
    System.halt(1)
  end

  defp perform(grid, action, {r1, c1, r2, c2}) do
    Enum.reduce(for(row <- r1..r2, col <- c1..c2, do: {row, col}), grid, fn
      coord, curr_grid ->
        case action do
          "turn on" ->
            current = Map.get(curr_grid, coord, 0)
            Map.put(curr_grid, coord, current + 1)
          "turn off" ->
            current = Map.get(curr_grid, coord, 0)
            Map.put(curr_grid, coord, if current != 0 do current - 1 else 0 end)
          "toggle" ->
            current = Map.get(curr_grid, coord, 0)
            Map.put(curr_grid, coord, current + 2)
        end
    end)
  end

  defp process(filename) do
    case File.read(filename) do
      {:ok, content} ->
        re = ~r/(turn on|turn off|toggle) (\d+),(\d+) through (\d+),(\d+)/
        grid =
          content
          |> String.split(~r/\r?\n/)
          |> Enum.reduce(%{}, fn
            line, curr_grid ->
              case Regex.run(re, line) do
                [_, action, r1s, c1s, r2s, c2s] ->
                  r1 = String.to_integer(r1s)
                  c1 = String.to_integer(c1s)
                  r2 = String.to_integer(r2s)
                  c2 = String.to_integer(c2s)
                  perform(curr_grid, action, {r1, c1, r2, c2})
                _ ->
                  curr_grid
              end
          end)
        total = Enum.sum(Map.values(grid))
        {:ok, total}
      {:error, reason} ->
        {:error, reason}
    end
  end
end

Program.main(System.argv())

File: scala/2015/day06b/Program.scala
import scala.io.Source
import scala.util.matching.Regex
import scala.util.Using
import scala.sys.exit

object Program {
  def usage(): Unit = {
    Console.err.println("usage: scala Program <input file>")
    sys.exit(1)
  }

  def perform(grid: Array[Int], action: String, r1: Int, c1: Int, r2: Int, c2: Int) = {
    for (row <- r1 to r2) {
      for (col <- c1 to c2) {
        action match {
          case "turn on" => grid(row * 1000 + col) += 1
          case "turn off" => grid(row * 1000 + col) = math.max(0, grid(row * 1000 + col) - 1)
          case "toggle" => grid(row * 1000 + col) += 2
          case _ => ()
        }
      }
    }
  }

  def process(filename: String): Int = {
    val re: Regex = raw"(turn on|turn off|toggle) (\d+),(\d+) through (\d+),(\d+)".r
    val grid: Array[Int] = Array.fill(1000*1000)(0)
    Using(Source.fromFile(filename)) { source =>
      source.getLines().foreach({ line =>
        re.findFirstMatchIn(line) match {
          case Some(matches) =>
             val action = matches.group(1)
             val r1 = matches.group(2).toInt
             val c1 = matches.group(3).toInt
             val r2 = matches.group(4).toInt
             val c2 = matches.group(5).toInt
             perform(grid, action, r1, c1, r2, c2)
          case None => ()
        }
      })
    }
    grid.sum
  }

  def main(args: Array[String]): Unit = {
    args.toList match {
      case filename :: Nil =>
        val result = process(filename)
        println(s"result = $result")
      case _ =>
        usage()
    }
  }
}
